crates/rspack_loader_lightningcss/src/lib.rs

```diff
@@ -0,0 +1,215 @@
+
+#[derive(Derivative)]
+#[derivative(Debug)]
+pub struct LightningCssLoader {
+  id: Identifier,
+  #[derivative(Debug = "ignore")]
+  visitors: Option<Mutex<Vec<LightningcssLoaderVisitor>>>,
+  config: Config,
+}
+
+impl LightningCssLoader {
+  pub fn new(visitors: Option<Vec<LightningcssLoaderVisitor>>, config: Config) -> Self {
+    Self {
+      id: LIGHTNINGCSS_LOADER_IDENTIFIER.into(),
+      visitors: visitors.map(|v| Mutex::new(v)),
+      config,
+    }
+  }
+
+  async fn loader_impl(&self, loader_context: &mut LoaderContext<RunnerContext>) -> Result<()> {
+    let Some(resource_path) = loader_context.resource_path() else {
+      return Ok(());
+    };
+
+    let filename = resource_path.to_string_lossy().into_owned();
+
+    let Some(content) = std::mem::take(&mut loader_context.content) else {
+      return Ok(());
+    };
+
+    let content_str = match &content {
+      rspack_core::Content::String(s) => Cow::Borrowed(s.as_str()),
+      rspack_core::Content::Buffer(buf) => String::from_utf8_lossy(buf),
+    };
+
+    let mut parser_flags = ParserFlags::empty();
+    parser_flags.set(
+      ParserFlags::CUSTOM_MEDIA,
+      matches!(&self.config.draft, Some(draft) if draft.custom_media),
+    );
+    parser_flags.set(
+      ParserFlags::DEEP_SELECTOR_COMBINATOR,
+      matches!(&self.config.non_standard, Some(non_standard) if non_standard.deep_selector_combinator),
+    );
+
+    let option = ParserOptions {
+      filename: filename.clone(),
+      css_modules: None,
+      source_index: 0,
+      error_recovery: self.config.error_recovery.unwrap_or(true),
+      warnings: None,
+      flags: parser_flags,
+    };
+    let stylesheet = StyleSheet::parse(&content_str, option.clone())
+      .map_err(|e| rspack_error::error!(e.to_string()))?;
+
+    let mut stylesheet = to_static(
+      stylesheet,
+      ParserOptions {
+        filename: filename.clone(),
+        css_modules: None,
+        source_index: 0,
+        error_recovery: true,
+        warnings: None,
+        flags: ParserFlags::empty(),
+      },
+    );
+
+    if let Some(visitors) = &self.visitors {
+      let visitors = visitors.lock().await;
+      for v in visitors.iter() {
+        v(&mut stylesheet);
+      }
+    }
+
+    let targets = Targets {
+      browsers: self.config.targets,
+      include: self
+        .config
+        .include
+        .as_ref()
+        .map(|include| Features::from_bits_truncate(*include))
+        .unwrap_or(Features::empty()),
+      exclude: self
+        .config
+        .exclude
+        .as_ref()
+        .map(|exclude| Features::from_bits_truncate(*exclude))
+        .unwrap_or(Features::empty()),
+    };
+
+    let unused_symbols = self
+      .config
+      .unused_symbols
+      .clone()
+      .map(|unused_symbols| unused_symbols.into_iter().collect())
+      .unwrap_or_default();
+
+    stylesheet
+      .minify(MinifyOptions {
```

Comment:
Do we need to minify the CSS code via this loader? I thought it should be done by `LightningCssMinimizerRspackPlugin`

Reply:
The LightningCssMinimizerRspackPlugin provides the `unusedSymbols` option too, maybe it is better to move symbols in the minification stage?

Reply:
the minify API is for transform here, lightningcss `transform` API also uses `minify` method